[ci skip] Fix the API docs for Bi-directional associations

repinel opened the pull request:
The examples now take in consideration when [Active Record finds inverse
associations automatically](https://github.com/rails/rails/pull/9522).

Fix #20528

zzak commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -955,35 +955,53 @@ def association_instance_set(name, association)
     # is the +evil_wizard+ association on +Dungeon+ (and vice-versa). By default,
-    # Active Record doesn't know anything about these inverse relationships and so no object
-    # loading optimization is possible. For example:
+    # Active Record uses an automatic guessing of the inverse association based on the name
```
"Active Record can guess the inverse of the association based on the name.."

zzak commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -955,35 +955,53 @@ def association_instance_set(name, association)
     #    d.level == t.dungeon.level # => true
     #    d.level = 10
-    #    d.level == t.dungeon.level # => false
+    #    d.level == t.dungeon.level # => true
```
This example doesn't make sense to me..

zzak commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -955,35 +955,53 @@ def association_instance_set(name, association)
     # The +Dungeon+ instances +d+ and <tt>t.dungeon</tt> in the above example refer to
-    # the same object data from the database, but are actually different in-memory copies
-    # of that data. Specifying the <tt>:inverse_of</tt> option on associations lets you tell
+    # the same in-memory instance thanks the heuristic applied. For this similar case:
```
Not sure this is an improvement.

zzak commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -955,35 +955,53 @@ def association_instance_set(name, association)
     #
     #    class Trap < ActiveRecord::Base
-    #      belongs_to :dungeon, inverse_of: :traps
+    #      belongs_to :special_dungeon, class_name: 'Dungeon', foreign_key: :dungeon_id, inverse_of: :traps
```
Doesn't this complicate the example?

zzak commented on activerecord/lib/active_record/associations.rb:
```diff
@@ -955,35 +955,53 @@ def association_instance_set(name, association)
     #    end
     #
-    # Then, from our code snippet above, +d+ and <tt>t.dungeon</tt> are actually the same
+    # Then, from the same snippet, +d+ and <tt>t.dungeon</tt> are actually the same
```
If we're going to show the code above, this should stay the same.

repinel committed:
[ci skip] Fix the API docs for Bi-directional associations

The examples now take in consideration when Active Record finds inverse
associations automatically.

repinel commented:
@zzak Thanks for the feedback!

I shrunk the changes to avoid over complicating the docs. I still have the example in a similar way, so let me know if it's clear.

zzak commented:
@repinel This is much more clear, thank you!!

repinel commented:
Thanks :heart:
